packages/web/src/components/twitter-auth/TwitterAuth.tsx

```diff
@@ -0,0 +1,173 @@
+import { CSSProperties, MouseEventHandler, ReactNode } from 'react'
+
+import 'whatwg-fetch'
+import 'url-search-params-polyfill'
+
+export type TwitterAuthProps = {
+  children?: ReactNode
+  className?: string
+  credentials?: 'omit' | 'same-origin' | 'include'
+  customHeaders?: Record<string, any>
+  dialogHeight?: number
+  dialogWidth?: number
+  forceLogin?: boolean
+  loginUrl: string
+  onClick: () => void
+  onFailure: (error: any) => void
+  onSuccess: (response: Body) => void
+  requestTokenUrl: string
+  screenName?: string
+  style?: CSSProperties
+  text?: string
+}
+
+const messages = {
+  signIn: 'Sign in with Twitter'
+}
+
+const TwitterAuth = (props: TwitterAuthProps) => {
+  const {
+    children,
+    className,
+    credentials = 'same-origin',
+    customHeaders,
+    dialogHeight = 400,
+    dialogWidth = 600,
+    forceLogin = false,
+    loginUrl,
+    onClick,
+    onFailure,
+    onSuccess,
+    requestTokenUrl,
+    screenName,
+    style,
+    text = messages.signIn
+  } = props
+
+  const onButtonClick: MouseEventHandler<HTMLDivElement> = (e) => {
+    e.preventDefault()
+    onClick?.()
+    return getRequestToken()
+  }
+
+  const getHeaders = () => {
+    return { ...customHeaders, 'Content-Type': 'application/json' }
+  }
+
+  const getRequestToken = () => {
+    const popup = openPopup()
+
+    return window
+      .fetch(requestTokenUrl, {
+        method: 'POST',
+        credentials,
+        headers: getHeaders()
+      })
+      .then((response) => {
+        return response.json()
+      })
+      .then((data) => {
+        let authenticationUrl = `https://api.twitter.com/oauth/authenticate?oauth_token=${data.oauth_token}&force_login=${forceLogin}`
+
+        if (screenName) {
+          authenticationUrl = `${authenticationUrl}&screen_name=${screenName}`
+        }
+
+        if (popup) {
+          popup.location = authenticationUrl
+          polling(popup)
+        }
+      })
+      .catch((error) => {
+        popup?.close()
+        return onFailure(error)
+      })
+  }
+
+  const openPopup = () => {
+    const w = dialogWidth
+    const h = dialogHeight
+
+    return window.open(
+      '',
+      '',
+      `toolbar=no, location=no, directories=no, status=no, menubar=no, scrollbars=no, resizable=no, copyhistory=no, width=${w}, height=${h}`
+    )
+  }
+
+  const polling = (popup: Window) => {
+    const polling = setInterval(() => {
```

Comment:
lil confusing with same name for outer fn

Reply:
yea also not using the interval id for anything